File: apps/mobile/composeApp/src/commonMain/kotlin/media/conduit/mobile/AudioTrackDisplay.kt
package media.conduit.mobile

import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

data class AudioTrackDisplayInfo(
    val title: String,
    val languageCode: String? = null,
    val languageName: String,
    val codec: String? = null,
    val channels: String? = null,
    val channelCount: Int? = null,
    val sampleRate: Int? = null,
    val bitrate: Long? = null,
)

data class AudioTrackDisplay(
    val primary: String,
    val secondary: String,
)

fun audioTrackDisplay(info: AudioTrackDisplayInfo, fallback: String): AudioTrackDisplay {
    val value = coreValue(buildJsonObject {
        put("type", "audioTrackDisplay")
        put("info", buildJsonObject {
            put("title", info.title)
            put("languageName", info.languageName)
            put("codec", info.codec)
            put("channels", info.channels)
            put("channelCount", info.channelCount)
            put("sampleRate", info.sampleRate)
            put("bitrate", info.bitrate)
        })
        put("fallback", fallback)
    }).jsonObject
    return AudioTrackDisplay(
        primary = value.getValue("primary").jsonPrimitive.content,
        secondary = value.getValue("secondary").jsonPrimitive.content,
    )
}

File: apps/web/src/lib/audio-track-display.ts
import type { NativeTrack } from "./desktop"
import { coreValue } from "./core"
import { trackLanguageName } from "./track-display"

export interface AudioTrackDisplay {
  primary: string
  secondary: string
}

export function audioTrackDisplay(track: NativeTrack, fallback: string): AudioTrackDisplay {
  const language = trackLanguageName(track.lang) ?? track.lang ?? "Unknown language"
  return coreValue<AudioTrackDisplay>({
    type: "audioTrackDisplay",
    info: {
      title: track.title ?? "",
      languageName: language,
      codec: track.codec,
      channels: track.audioChannels,
      channelCount: track.channelCount,
      sampleRate: track.sampleRate,
      bitrate: track.bitrate,
    },
    fallback,
  })
}
